### PR TITLE
GPT-J support reuse_cache

### PR DESCRIPTION
**Pull request by atakaha**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by atakaha** [comment]
RoPE and FusedSDPA enabling will be committed by other PRs.

**Comment by libinta** [comment]
@atakaha  can you provide performance for below and adding test case? 

Model | Batch Size | TP (1) | Input Length | Output Length |
-- | -- | -- | -- | -- | --
GPT-J 6B | 512 | 1 | 128 | 128 |
GPT-J 6B | 32 | 1 | 128 | 2048 | 
GPT-J 6B | 32 | 1 | 2048 | 128 | 
GPT-J 6B | 16 | 1 | 2048 | 2048 |

**Comment by atakaha** [comment]
@libinta, Here is performance table. All enabled reuse_cache

| Model | Batch Size | Input Length | Output Length | precision | Throughtput (tokens/sec) | Mem Alloc (GB) | Max Mem alloc (GB) |
| --- | --- | --- | --- | --- | --- | --- | --- | 
| GPT-J 6B | 512 | 128 | 128 | bf16 | 8076.863 | 73.87 | 88.08 |
| GPT-J 6B | 512 | 128 | 128 | fp8 | 14694.365 | 40.42 | 66.34 |
| GPT-J 6B | 32 | 128 | 2048 | bf16 | 1561.541 | 41.34 | 42.71 |
| GPT-J 6B | 32 | 128 | 2048 | fp8 | 2859.099 | 21.04 | 21.88 |
| GPT-J 6B | 32 | 2048 | 128 | bf16 | 765.932 | 49.63 | 76.06 |
| GPT-J 6B | 32 | 2048 | 128 | fp8 | 1257.655 | 26.9 | 52.82 |
| GPT-J 6B | 16 | 2048 | 2048 | bf16 | 799.621 | 43.58 | 56.59 |
| GPT-J 6B | 16 | 2048 | 2048 | fp8 | 1450.878 | 22.92 | 35.68 |

**Comment by yafshar** [comment]
@atakaha the 

> @libinta, Here is performance table. All enabled reuse_cache
> Model 	Batch Size 	Input Length 	Output Length 	precision 	Throughtput (tokens/sec) 	Mem Alloc (GB) 	Max Mem alloc (GB)
> GPT-J 6B 	512 	128 	128 	bf16 	8076.863 	73.87 	88.08
> GPT-J 6B 	512 	128 	128 	fp8 	14694.365 	40.42 	66.34
> GPT-J 6B 	32 	128 	2048 	bf16 	1561.541 	41.34 	42.71
> GPT-J 6B 	32 	128 	2048 	fp8 	2859.099 	21.04 	21.88
> GPT-J 6B 	32 	2048 	128 	bf16 	765.932 	49.63 	76.06
> GPT-J 6B 	32 	2048 	128 	fp8 	1257.655 	26.9 	52.82
> GPT-J 6B 	16 	2048 	2048 	bf16 	799.621 	43.58 	56.59
> GPT-J 6B 	16 	2048 	2048 	fp8 	1450.878 	22.92 	35.68

@atakaha how did you run for this performance table. I see some mismatch. For example, GPT-J 6B, BS=16 INN/OUT=2048 uses more that 80GB memory. Would you provide me with your command.

**Comment by atakaha** [comment]
@yafshar 
The sample command lines I used are:
- measurement
QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_generation.py --model_name_or_path EleutherAI/gpt-j-6b --use_hpu_graphs --use_kv_cache --reuse_cache --limit_hpu_graphs --max_input_tokens 128 --max_new_tokens 128 --batch_size 1 --bf16

-  bf16
python run_generation.py --model_name_or_path EleutherAI/gpt-j-6b --use_hpu_graphs --use_kv_cache --reuse_cache --limit_hpu_graphs --max_input_tokens 2048 --max_new_tokens 2048 --batch_size 16 --bf16

- fp8
QUANT_CONFIG=quantization_config/maxabs_quant.json python run_generation.py --model_name_or_path EleutherAI/gpt-j-6b --use_hpu_graphs --use_kv_cache --reuse_cache --limit_hpu_graphs --max_input_tokens 2048 --max_new_tokens 2048 --batch_size 16 --bf16

**Comment by libinta** [comment]
Seems bf16 2048/2048 are not as good as A100 fp16 882 TensorRT-LLM/docs/source/performance/perf-overview.md at main * NVIDIA/TensorRT-LLM (github.com)<https://github.com/NVIDIA/TensorRT-LLM/blob/main/docs/source/performance/perf-overview.md>, fp8 is only half of H100.
There are more optimizations can be ported, please check other models such as llama, falcon, mistral, qwen2.

**Comment by atakaha** [comment]
for more optimization, I'm going to add FusedSDPA and others after this.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 1]
```diff
@@ -1,20 +1,116 @@
...
+    GPTJBlock,
     GPTJForCausalLM,
+    GPTJModel,
     apply_rotary_pos_emb,
     create_sinusoidal_positions,
     logger,
 )
 
 
+class Matmul(nn.Module):
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:21]: @atakaha the `nn.Module` wrapper for FP8 is no longer needed, use the `torch.matmul` directly in the code. [torch.matmul and torch.nn.functional.softmax are supported and replaced by HQT during quantization.](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html?highlight=fp8%20#supported-functions)

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 2]
```diff
@@ -31,7 +127,7 @@ def _attn(
         key = key.contiguous()
         value = value.contiguous()
 
-        attn_weights = torch.matmul(query, key.transpose(-1, -2))
+        attn_weights = self.matmul_qk(query, key.transpose(-1, -2))
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:130]: @atakaha above comment

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 3]
```diff
@@ -1,20 +1,116 @@
...
+        self.max_position_embeddings = config.max_position_embeddings
+
+    def allocate_kv_cache(self, batch_size, max_seq_len, inp_seq_len):
+        cache_shape = (batch_size, self.num_attention_heads, max_seq_len, self.head_dim)
+        device = self.k_proj.weight.device
+        dtype = self.config.torch_dtype
+        self.k_cache.allocate(inp_seq_len, dtype, device, cache_shape)
+        self.v_cache.allocate(inp_seq_len, dtype, device, cache_shape)
+
+    def reorder(self, tensor, beam_idx, dim_a, dim_b):
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:89]:
```suggestion
    def reorder(self, tensor, beam_idx):
```
Is there a reason for these extra arguments?

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 4]
```diff
@@ -1,20 +1,116 @@
...
+        tensor.copy_(updated)
+
+    def reorder_kv_cache(self, beam_idx: torch.LongTensor):
+        if self.k_cache.cache is None:
+            return (None, None)
+
+        head_dim = self.k_cache.cache.size(-1)
+        seq_length = self.k_cache.cache.size(-2)
+        self.reorder(self.k_cache.cache, beam_idx, seq_length, head_dim)
+        self.reorder(self.v_cache.cache, beam_idx, seq_length, head_dim)
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:100]:
```suggestion
        self.reorder(self.k_cache.cache, beam_idx)
        self.reorder(self.v_cache.cache, beam_idx)
```

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 5]
```diff
@@ -196,202 +333,226 @@ def forward(
 
         return outputs  # hidden_states, present, (attentions)
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:335]:
```suggestion


```

Add one extra empty line. 2 lines between classes!

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 6]
```diff
@@ -196,202 +333,226 @@ def forward(
...
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        token_idx: Optional[torch.Tensor] = None,
+        reuse_cache: Optional[bool] = False,
+        cache_idx: Optional[int] = None,
+        sin: Optional[torch.Tensor] = None,
+        cos: Optional[torch.Tensor] = None,
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:369]:
```suggestion
```
I am a bit confused! which upper level these args are coming from. You are overwriting them later

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 7]
```diff
@@ -77,10 +175,13 @@ def forward(
         Copied from GPTJAttention.forward: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:175]: @atakaha there is mismatch from the information. Here, you mentioned that you copied this piece of code from transformers master branch. At line 267, it is copied from 4.38.2. Please be consistent!

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 8]
```diff
@@ -1,20 +1,116 @@
...
+    def forward(self, cur, dim, idx):
+        return self.update(self.cache, cur, dim, idx, self.inp_seq_len)
+
+
 class GaudiGPTJAttention(GPTJAttention):
+    def __init__(self, config: GPTJConfig):
+        super().__init__(config)
+        self.config = config
+
+        self.matmul_qk = Matmul()
```
[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:75]: same above.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 2 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:129]: same above.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 3 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: done.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 4 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: done.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 5 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:334]: done.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 6 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:369]: These are original code.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 7 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: This is from original and updated.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 6 above)
Comment (yafshar) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:369]: I do not see any `sin` and `cos` in the original code's forward interface. Can you post the link to that. Even if it is the case st is not right!

Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:369]: do you mean this?
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/gptj/modeling_gptj.py#L214

Comment (yafshar) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:369]: Oh, thanks, you got this code from optimum-habana not transformers. I do not see what is the point of having this argument in the interface unless there is justification for that or you correct the logic inside? 

@regisss your input will help here

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 9]
```diff
@@ -121,10 +221,29 @@ def forward(
...
+                present = layer_past
+
+            if cache_idx is not None and q_len == 1:
+                key = key[:, :, :cache_idx, :]
+                value = value[:, :, :cache_idx, :]
+                if attention_mask is not None:
+                    attention_mask = attention_mask[:, :, :, :cache_idx]
+
             # Note that this cast is quite ugly, but is not implemented before ROPE as the original codebase keeps the key in float32 all along the computation.
             # Reference: https://github.com/kingoflolz/mesh-transformer-jax/blob/f8315e3003033b23f21d78361b288953064e0e76/mesh_transformer/layers.py#L128
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:245]:
```suggestion
```
Can we remove this? It is not relevant anymore

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 10]
```diff
@@ -121,10 +221,29 @@ def forward(
...
-            present = (key.to(hidden_states.dtype), value)
+            # present = (key.to(hidden_states.dtype), value)
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:246]:
```suggestion
```
The same as above

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 11]
```diff
@@ -197,201 +333,226 @@ def forward(
...
+        sincos = torch.gather(embed_positions, 1, repeated_position_ids)
+        sin, cos = torch.split(sincos, sincos.shape[-1] // 2, dim=-1)
+        sin = sin.contiguous()
+        cos = cos.contiguous()
+
+        htcore.mark_step()
+        for i, (block, layer_past) in enumerate(zip(self.h, past_key_values)):
+            # Model parallel
+            if self.model_parallel:
+                torch.cuda.set_device(hidden_states.device)
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:484]:
```suggestion
                torch.hpu.set_device(hidden_states.device)
```

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 12]
```diff
@@ -197,201 +333,226 @@ def forward(
...
+            if use_cache is True:
+                presents = presents + (outputs[1],)
+
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (outputs[2 if use_cache else 1],)
+
+            # Model Parallel: If it's the last layer for that device, put things on the next device
+            if self.model_parallel:
+                for k, v in self.device_map.items():
+                    if i == v[-1] and "cuda:" + str(k) != self.last_device:
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/modeling_gptj.py:537]: @atakaha Have you done any test on model_parallel? We should fix this logic!

Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:537]: this is original code and didn't tested model_parallel, since scope of this is 1 TP.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py [hunk 13]
```diff
@@ -197,201 +333,226 @@ def forward(
...
+                        hidden_states = hidden_states.to("cuda:" + str(k + 1))
```
[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:538]: this is original code.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 11 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:484]: this is original code.

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 9 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: done

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 10 above)
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: done

File: optimum/habana/transformers/models/gptj/__init__.py [hunk 14]
```diff
@@ -1,6 +1,7 @@
+from transformers.models.gptj.configuration_gptj import GPTJConfig
```
Comment (yafshar) [inline optimum/habana/transformers/models/gptj/__init__.py:1]:
```suggestion
from transformers.models.gptj.configuration_gptj import GPTJConfig

```

**Review by yafshar — APPROVED** [review]
Look good for now!

@atakaha will follow up with more optimizations in the separate PRs to be comparative. 
@regisss would you please check this! Since the performance is not comparative, please don't put into front page! Thanks

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 6 above)
Comment (regisss) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:365]: Yes, I agree it looks strange. It comes from https://github.com/huggingface/optimum-habana/pull/468.
@BaihuiJin Do you remember why you added that?

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 11 above)
Comment (regisss) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:480]: We can remove of the code blocks starting with `if self.model_parallel` actually, it's not going to be used on HPU

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 12 above)
Comment (regisss) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:533]: Same you can remove it. We would do TP differently on HPU I guess.

Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: removed the block

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: removed the block

File: optimum/habana/transformers/models/gptj/modeling_gptj.py (hunk 6 above)
Comment (BaihuiJin) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py:365]: Oh, I was meant to pass these arguments to the `gaudi_gptj_block_forward`, as previously sin and cos was generated in in each blocks, which could be optimized. In the first design I tried to acquire sin and cos in the upper level and pass to `forward`, hence the arguments. But in the final design I changed to generate sincos in `forward` method. I must have forgotten to remove these arguments. It's a mistake, sorry. I think its ok to remove them.

Comment (atakaha) [reply optimum/habana/transformers/models/gptj/modeling_gptj.py]: removed these.

**Review by regisss — APPROVED** [review]
LGTM
